File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/userScanReceipt/UserScanReceiptHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.userScanReceipt

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.userScanReceipt.UserScanReceiptApi
import features.userScanReceipt.models.ScanReceiptResponseModel

import features.userScanReceipt.serializeUserScanReceiptState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import features.userScanReceipt.models.ScanReceiptRequest

class UserScanReceiptHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/scan_receipt"

    override val eventChannel: String
        get() = "customer_sdk/events/scan_receipt"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return scanReceipt(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun scanReceipt(call: MethodCall): String? {
            val result = UserScanReceiptApi().scanReceipt(
                scanReceiptRequest = Json.decodeFromString(call.argument<String>("scanReceiptRequest") as String)
            ).toJsonResult<ScanReceiptResponseModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        UserScanReceiptApi().userScanReceiptState.serializeUserScanReceiptState().observe { result ->
            pushSerializedResultToEventSink<UserScanReceiptApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_macos/macos/Classes/features/userScanReceipt/UserScanReceiptHandler.swift
import Foundation
import FlutterMacOS
import KangarooAppSdkCustomer

class UserScanReceiptHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/scan_receipt"

    var eventChannel: String = "customer_sdk/events/scan_receipt"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await UserScanReceiptHandler.scanReceipt(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func scanReceipt(call: FlutterMethodCall) async -> String? {


        
    let args = call.arguments
        do {

        var overrideHeaders: [String: String]?

        if let myArgs = args as? [String: Any] {
            overrideHeaders = myArgs["overrideHeaders"] as? [String: String]
        }
        else {
            overrideHeaders = [:]
        }

        let result = try await UserScanReceiptApi().scanReceipt(overrideHeaders: overrideHeaders, methods: call.arguments as! [String : Any]).serializeUserScanReceiptApiResult()

            switch result {
                case let result as SerializedResultSuccess:
                    return result.data
                case let result as SerializedResultUnauthorizedError:
                    return result.error
                case let result as SerializedResultUnknownError:
                    return result.error
                default:
                    return nil
            }

        }
        catch {
            return nil
        }


        


        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        UserScanReceiptApi().userScanReceiptState
            .serializeUserScanReceiptState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
